### PR TITLE
GH-94979: Pop the exception stack when jumping out of exception handlers.

### PR DESCRIPTION
**Pull request by markshannon**
Skipping news as there have been no releases between the introduction of the bug and the fix.

* Issue: gh-94979

### CODE REVIEW

File: Objects/frameobject.c [hunk 1]
```diff
@@ -782,7 +812,18 @@ frame_setlineno(PyFrameObject *f, PyObject* p_new_lineno, void *Py_UNUSED(ignore
         start_stack = pop_value(start_stack);
     }
     while (start_stack > best_stack) {
-        frame_stack_pop(f);
+        if (top_of_stack(start_stack) == Except) {
+            /* Pop exception stack as well as the evaluation stack */
+            PyThreadState *tstate = _PyThreadState_GET();
+            _PyErr_StackItem *exc_info = tstate->exc_info;
+            PyObject *value = exc_info->exc_value;
+            exc_info->exc_value = _PyFrame_StackPop(f->f_frame);
```
Comment (iritkatriel) [inline Objects/frameobject.c]: Should we assert that it's an exception?

**Comment by markshannon** [comment]
1 buildbot seems to have stuck running `test_concurrent_futures`.

**Review by iritkatriel — APPROVED** [review]
FTR: this is getting crazy.

**Comment by markshannon** [comment]
> FTR: this is getting crazy.

I assume you mean the duplication of stack effect here, in the compiler and the interpreter.
That's why we want to build all three from a single declarative form, if we can.

**Pull request merged**